### DISCUSSION TITLE
x-for over JSON data (v3)

### DISCUSSION THREAD

**Discussion by jeffersonbenson** [alpinejs/alpine#3228] [1. Help] [answered]
Hey all!

I'm fairly new to full-stack web development and Alpine, but I fell in love with the framework once I watched the overview and didn't feel over my head. I've got a project loaded up that pulls JSON data from an express server I built and attempts to display it as a card.

However, while I'm able to view the data in `console.log()` I'm not seeing it in the AlpineJS Dev Tools plugin.
![image](https://user-images.githubusercontent.com/19256220/197654899-b6a71d55-d2e5-4391-ba5e-2430145fe9f0.png)
![image](https://user-images.githubusercontent.com/19256220/197654923-51d10c55-5d25-4ee9-88da-3e922ffe6e20.png)

I'm pretty sure I'm doing something obviously wrong, but I realized that a lot of the documentation available is for v2 which isn't the most helpful. Any insight would be greatly appreciated. Thanks!

### Goal: The Express server produces an array of objects when a given key is passed. The frontend then produces a card for each item in the array with the name and docker image in use. 

```javascript
<script>
    async function connect(){
      try {
          pubKey = resp.publicKey.toString();
          this.shortpubKey = pubKey.substring(0,7) + '...'
          const lowerpubKey = pubKey.toLowerCase();
          pods = await getPods(lowerpubKey);
          console.log(pods);
      }catch (err) {
          this.shortpubKey = 'Error!'
          console.log("Something has gone wrong!")
          console.log(err)
      }
    }
    async function getPods(lowerpubKey) {
        let response = await fetch('http://localhost:8080/getpods/' + lowerpubKey);
        let pods = await response.json();
        return pods;
    }
    </script>
```

```html
<div class="columns mt-5 is-8 is-variable" x-data="{ pods : this.pods }">

          <template x-for="pod in this.pods" :key="pod">
          <div class="column is-4-tablet is-3-desktop">
            <div class="card">
              <div class="card-content">
                <p class="title has-text-primary-dark" x-text="pod.name"></p>
                <p class="subtitle" x-text="pod.image"></p>
                <p>
                  <a href="" target="_blank" rel="noopener noreferrer">mytest</a>
                </p>
                <p>
                  Time Remaining:
                </p>
              </div>
              <footer class="card-footer">
                <a class="card-footer-item" href="">Logs</a>
                <a class="card-footer-item" href="">Add Time</a>
              </footer>
            </div>
          </div>
          </template>

        </div>
```

**Comment by ekwoka** [comment]
You don't need to use `this.` in expressions in the way you are there.

It also doesn't seem that you ever actually add the pods to your component.

**Reply by jeffersonbenson** [reply]
@ekwoka Thanks for the feedback. I removed the `this.` and am seeing the following error in my browser. I think it may be related to what you mean to add the pods to the component, but could you elaborate a bit further what you mean by "adding to the component"?
![image](https://user-images.githubusercontent.com/19256220/198331622-32b94624-59e7-4742-bec4-5941a93bcafa.png)

**Reply by ekwoka** [reply]
Well, where do you define "pods"?

**Reply by jeffersonbenson** [reply]
Just in the <script> tag above. I do the same thing for the shortpubkey variable.

For some context, here's a link to the full repo: https://github.com/jeffersonbenson/starship

**Reply by ekwoka** [reply]
Okay, I see it.

Yeah, so pods is undefined until after you connect wallet, but you're trying to access `pods` before that.

Instead, try running the `getPods` function inside the component to assign pods to the Alpine component.

Like the x-data can have an `init` method that does the get pods.

**Reply by jeffersonbenson** [reply]
I think I understand what you mean. Will the data be reactive if I load `x-data` with `getPods`? `getPods` shouldn't have any data until after `connectWallet` runs, so I'm not sure an `init` method would work in this situation.

**Reply by ekwoka** [reply]
Then run connect wallet in the component, or have getPods wait until connect wallet runs.

Many different ways to structure it in a manner to work.

**Reply by jeffersonbenson** [reply]
Gotcha. I'll give it a shot and see what happens. Thanks!

**Reply by jeffersonbenson** [reply]
Okay, I really think. I'm so close to understanding what to do here. I refactored my `<script>` tag code to wait until `lowerpubKey` is not an empty string. 
```javascript
<script>
    // Solana code goes here
    const endpoint = solanaWeb3.clusterApiUrl('devnet');
    let lowerpubKey = '';
    let pods = {};
    async function connectWallet(){
      try {
          const resp = await window.solana.connect();
          pubKey = resp.publicKey.toString();
          this.shortpubKey = pubKey.substring(0,7) + '...'
          lowerpubKey = pubKey.toLowerCase();
          console.log("Connected to the server? "+window.solana.isConnected+" and your public key is "+pubKey);
          //pods = await getPods(lowerpubKey);
          //console.log(pods);
      }catch (err) {
          this.shortpubKey = 'Error!'
          console.log("Something has gone wrong!")
          console.log(err)
      }
    }
    async function getPods() {
      if ( lowerpubKey == '' ) {
        console.log('No lowerpubKey found!')
        window.setTimeout(getPods, 1000);
      }
      else {
        let response = await fetch('http://localhost:8080/getpods/' + lowerpubKey);
        let pods = await response.json();
        console.log(pods);
        return pods;
      }
    }
    </script>
```

I then set my x-data to the value of the function (I think).

```html
<div class="container">
    
        <div class="columns mt-5 is-8 is-variable" x-data="{ getPods }">

          <template x-for="pod in pods" :key="pod.id">
```

And I'm seeing this as the value for x-data:

![image](https://user-images.githubusercontent.com/19256220/198850712-fc9c8b82-29dc-48f6-9990-8ce79bd08d56.png)

Shouldn't it be reactive when `getPods()` runs? Or am I missing an `x-init` to first grab the blank value before filling it with data?

**Reply by SimoTod** [reply]
getPods is a function and you are not invoking it so the output is correct. Your function is async, so you need to use await but you can't really await inside x-data. Await will cause the function to pause and Alpine will start evaluating the inner directives triggering errors. The canonical way so far is to initiate the variable with an empty value and use x-init to instantiate it.
`x-data="{pods: []}" x-init="pods = await getPods()"`

**Reply by jeffersonbenson** [reply]
Thank you @SimoTod. So while your explanation makes sense, `pods` is now showing as 'undefined' in the Alpine JS Devtools. For some additional context, here's what the fetch'd JSON looks like:
![image](https://user-images.githubusercontent.com/19256220/198858298-88e3e5a4-6f10-44eb-86c6-79de854a1c0a.png)
Perhaps it's something in the return statement from the `<script>` tag?

**Reply by SimoTod** [reply]
I suppose it's the if else inside getPods. 

- Alpine starts
- X-init calls getPods
- getPods sets a timeout and doesn't return anything
- pods is undefined
- after a bit, the timeout triggers the function again, the function returns the pod and prints out the console.log but it's not assigned to anything

I'm not sure what are your constraints and how you thought your logic but it seems more complicate that it should be.

If the key is populated by an external process and you can't know when it's ready, you can't really rely on x-init. You can dispatch a custom event when the key is set and your alpine component can listen for it. The point is getPods must return something.

**Reply by jeffersonbenson** [reply]
So I would agree I believe I'm over-complicating the matter 😁 . I'll start researching into dispatching a custom event as a potential solution. 
Overall this is the logic I'm aiming for:
- User clicks on button
- Button calls `connectWallet()` and retrieves the user's public key via the Solana library
- `getPods()` makes a fetch request to a url containing the user's public key
- The JSON data returned is then displayed as a series of cards via the `<template>` element.

**Comment by jeffersonbenson** [comment, marked as answer]
Wanting to close out this discussion, but I finally figured out how to make this happen.
1. Initialize x-data in my component with an empty array
2. Have my Connect function fire off an event after the user's public key is known
3. Use that event to fetch the relevant data
4. Push the data into the empty array inside x-data using a spread operator (Also, TIL what a spread operator is 😁)

Here's the resulting line: 
![image](https://user-images.githubusercontent.com/19256220/201544062-c81f32e7-0744-457d-806a-73821ce5b053.png)

Many thanks to everyone who helped along the way. I think I'm finally starting to understand some of the fundamentals of Alpine in practice.
